include/ghex/rma/range_factory.hpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2020, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#ifndef INCLUDED_GHEX_RMA_RANGE_FACTORY_HPP
+#define INCLUDED_GHEX_RMA_RANGE_FACTORY_HPP
+
+#include <vector>
+#include <cstring>
+#include <boost/mp11.hpp>
+#include "./locality.hpp"
+#include "./handle.hpp"
+#include "./access_guard.hpp"
+#include "./range.hpp"
+
+namespace gridtools {
+namespace ghex {
+namespace rma {
+
+/** @brief Serializes and de-serializes range types which are among the types in RangeList. This
+  * class manages the type erasure and type injection before and after serialization.
+  * @tparam RangeList a list of range types */
+template<typename RangeList>
+struct range_factory
+{
+    using range_type = range;
+
+    static_assert(boost::mp11::mp_is_set<RangeList>::value, "range types must be unique");
+
+    static inline constexpr unsigned int a16(unsigned int size) noexcept { return ((size+15)/16)*16; }
+    
+    template<typename Range>
+    using range_size_p = boost::mp11::mp_size_t<sizeof(Range)>;
+    using max_range_size =
+        boost::mp11::mp_max_element<boost::mp11::mp_transform<range_size_p, RangeList>, boost::mp11::mp_less>;
+    
+    static constexpr std::size_t serial_size =a16(sizeof(int)) + a16(sizeof(info)) 
+        + a16(sizeof(typename local_access_guard::info)) + max_range_size::value;
+
+    template<typename Range>
+    static void serialize(info field_info, local_access_guard& g, const Range& r, unsigned char* buffer)
+    {
+        static_assert(boost::mp11::mp_set_contains<RangeList, Range>::value, "range type not registered");
+        using id = boost::mp11::mp_find<RangeList, Range>;
+        const int m_id = id::value;
+        std::memcpy(buffer, &m_id, sizeof(int));
+        buffer += a16(sizeof(int));
+        std::memcpy(buffer, &field_info, sizeof(field_info));
+        buffer += a16(sizeof(field_info));
+        auto info_ = g.get_info();
+        std::memcpy(buffer, &info_, sizeof(typename local_access_guard::info));
+        buffer += a16(sizeof(typename local_access_guard::info));
+        std::memcpy(buffer, &r, sizeof(Range));
+    }
+
+    template<typename Range>
+    static std::vector<unsigned char> serialize(info field_info, local_access_guard& g, const Range& r)
```

Comment:
why do you need this one? would the previous one be all you need?

Reply:
Right, for now one version is enough I think. I will change it.